### ISSUE TITLE
CUDA_ERROR_LAUNCH_TIMEOUT when using NCCL with MultiWorkerMirroredStrategy on multi-node systems

### ISSUE THREAD

**Issue by Flamefire** [tensorflow/tensorflow#37693]
**System information** 
- Have I written custom code (as opposed to using a stock
example script provided in TensorFlow): Based on https://www.tensorflow.org/tutorials/distribute/multi_worker_with_keras but using bigger dataset and Resnet
- OS Platform and Distribution (e.g.,Linux Ubuntu 16.04): RHEL 7
- TensorFlow installed from (source or binary): source
- TensorFlow version (use command below): 2.1.0
- Python version: 3.7.4
- Bazel version (if compiling from source): 0.29.1
- GCC/Compiler version (if compiling from source): GCC 8.3.0
- CUDA/cuDNN version: CUDA/10.1.243, cuDNN/7.6.4.38, NCCL/2.4.8
- GPU model and memory:Tesla K80

**Describe the current behavior**

Running the standard training loop results in a CUDA_ERROR_LAUNCH_TIMEOUT when using NCCL.

**Describe the expected behavior**

No error.

**Standalone code to reproduce the issue** 

```
import tensorflow as tf
import tensorflow_datasets as tfds
from slurm_utils import set_tf_config
from slurm_cluster_resolver import SlurmClusterResolver

tfds.disable_progress_bar()

BUFFER_SIZE = 10000
BATCH_SIZE = 32


def make_datasets_unbatched():
    def preprocess(image, label):
        image = tf.image.convert_image_dtype(image, tf.float32)
        image = tf.image.resize_with_pad(image, 299, 299)
        image = image - [123.68, 116.78, 103.94]
        return image, label

    datasets, info = tfds.load(name='stanford_dogs',
                               with_info=True,
                               as_supervised=True)

    return {
        'train':
        datasets['train'].map(preprocess).cache().shuffle(BUFFER_SIZE),
        'test': datasets['test'].map(preprocess)
    }, info


def build_and_compile_cnn_model(num_classes):
    model = tf.keras.applications.InceptionV3(weights=None,
                                              classes=num_classes)
    model.compile(loss=tf.keras.losses.sparse_categorical_crossentropy,
                  optimizer=tf.keras.optimizers.SGD(learning_rate=0.001),
                  metrics=['accuracy'])
    return model


resolver = SlurmClusterResolver()
set_tf_config(resolver)
strategy = tf.distribute.experimental.MultiWorkerMirroredStrategy(
    tf.distribute.experimental.CollectiveCommunication.NCCL,
    cluster_resolver=resolver)

# Here the batch size scales up by number of workers since
# `tf.data.Dataset.batch` expects the global batch size.
GLOBAL_BATCH_SIZE = BATCH_SIZE * strategy.num_replicas_in_sync
with strategy.scope():
    # Creation of dataset, and model building/compiling need to be within
    # `strategy.scope()`.
    options = tf.data.Options()
    options.experimental_distribute.auto_shard_policy = tf.data.experimental.AutoShardPolicy.DATA
    datasets, info = make_datasets_unbatched()
    for split in datasets:
        datasets[split] = datasets[split].batch(
            GLOBAL_BATCH_SIZE).with_options(options)
    model = build_and_compile_cnn_model(info.features['label'].num_classes)

train_steps = info.splits['train'].num_examples // GLOBAL_BATCH_SIZE
test_steps = info.splits['test'].num_examples // GLOBAL_BATCH_SIZE
print("Have %s examples -> %s batches of %s" %
      (info.splits['train'].num_examples, train_steps, GLOBAL_BATCH_SIZE))

model.fit(x=datasets['train'].repeat(),
          epochs=1,
          steps_per_epoch=train_steps,
          verbose=2)
```

Note: `SlurmClusterResolver` is the one that got recently included to TF and `set_tf_config` is:

```
def set_tf_config(cluster_resolver, environment=None):
    """Set the TF_CONFIG env variable from the given cluster resolver"""
    cfg = {
        'cluster': cluster_resolver.cluster_spec().as_dict(),
        'task': {
            'type': cluster_resolver.get_task_info()[0],
            'index': cluster_resolver.get_task_info()[1],
        },
        'rpc_layer': cluster_resolver.rpc_layer,
    }
    if environment:
        cfg['environment'] = environment
    os.environ['TF_CONFIG'] = json.dumps(cfg)
    print("Set TF_CONFIG=" + os.environ['TF_CONFIG'])
```

**Other info / logs**

This happens everytime when using more than 1 node.

```
Set TF_CONFIG={"cluster": {"worker": ["taurusi2107:8888", "taurusi2108:8888"]}, "task": {"type": "worker", "index": 1}, "rpc_layer": "grpc"}
Have 12000 examples -> 46 batches of 256
WARNING:tensorflow:`eval_fn` is not passed in. The `worker_fn` will be used if an "evaluator" task exists in the cluster.
WARNING:tensorflow:`eval_fn` is not passed in. The `worker_fn` will be used if an "evaluator" task exists in the cluster.
WARNING:tensorflow:`eval_strategy` is not passed in. No distribution strategy will be used for evaluation.
WARNING:tensorflow:`eval_strategy` is not passed in. No distribution strategy will be used for evaluation.
WARNING:tensorflow:ModelCheckpoint callback is not provided. Workers will need to restart training if any fails.
WARNING:tensorflow:ModelCheckpoint callback is not provided. Workers will need to restart training if any fails.
Train for 46 steps
Have 12000 examples -> 46 batches of 256
WARNING:tensorflow:`eval_fn` is not passed in. The `worker_fn` will be used if an "evaluator" task exists in the cluster.
WARNING:tensorflow:`eval_fn` is not passed in. The `worker_fn` will be used if an "evaluator" task exists in the cluster.
WARNING:tensorflow:`eval_strategy` is not passed in. No distribution strategy will be used for evaluation.
WARNING:tensorflow:`eval_strategy` is not passed in. No distribution strategy will be used for evaluation.
WARNING:tensorflow:ModelCheckpoint callback is not provided. Workers will need to restart training if any fails.
WARNING:tensorflow:ModelCheckpoint callback is not provided. Workers will need to restart training if any fails.
Train for 46 steps
2020-03-18 16:51:46.733598: E tensorflow/stream_executor/cuda/cuda_event.cc:29] Error polling for event status: failed to query event: CUDA_ERROR_LAUNCH_TIMEOUT: the launch timed out and was terminated
2020-03-18 16:51:46.733598: E tensorflow/stream_executor/cuda/cuda_event.cc:29] Error polling for event status: failed to query event: CUDA_ERROR_LAUNCH_TIMEOUT: the launch timed out and was terminated
2020-03-18 16:51:46.733599: E tensorflow/stream_executor/cuda/cuda_event.cc:29] Error polling for event status: failed to query event: CUDA_ERROR_LAUNCH_TIMEOUT: the launch timed out and was terminated
2020-03-18 16:51:46.733602: E tensorflow/stream_executor/cuda/cuda_event.cc:29] Error polling for event status: failed to query event: CUDA_ERROR_LAUNCH_TIMEOUT: the launch timed out and was terminated
2020-03-18 16:51:46.733657: F tensorflow/core/common_runtime/gpu/gpu_event_mgr.cc:273] Unexpected Event status: 1
2020-03-18 16:51:46.733656: F tensorflow/core/common_runtime/gpu/gpu_event_mgr.cc:273] Unexpected Event status: 1
2020-03-18 16:51:46.733662: F tensorflow/core/common_runtime/gpu/gpu_event_mgr.cc:273] Unexpected Event status: 1
2020-03-18 16:51:46.733671: F tensorflow/core/common_runtime/gpu/gpu_event_mgr.cc:273] Unexpected Event status: 1
[taurusi2108:15936] *** Process received signal ***
2020-03-18 16:51:47.705311: E tensorflow/stream_executor/cuda/cuda_event.cc:29] Error polling for event status: failed to query event: CUDA_ERROR_LAUNCH_TIMEOUT: the launch timed out and was terminated
2020-03-18 16:51:47.705311: E tensorflow/stream_executor/cuda/cuda_event.cc:29] Error polling for event status: failed to query event: CUDA_ERROR_LAUNCH_TIMEOUT: the launch timed out and was terminated
2020-03-18 16:51:47.705349: F tensorflow/core/common_runtime/gpu/gpu_event_mgr.cc:273] Unexpected Event status: 1
2020-03-18 16:51:47.705311: E tensorflow/stream_executor/cuda/cuda_event.cc:29] Error polling for event status: failed to query event: CUDA_ERROR_LAUNCH_TIMEOUT: the launch timed out and was terminated
2020-03-18 16:51:47.705359: F tensorflow/core/common_runtime/gpu/gpu_event_mgr.cc:273] Unexpected Event status: 1
[taurusi2107:31475] *** Process received signal ***
2020-03-18 16:51:47.705372: F tensorflow/core/common_runtime/gpu/gpu_event_mgr.cc:273] Unexpected Event status: 1
2020-03-18 16:51:47.705326: E tensorflow/stream_executor/cuda/cuda_driver.cc:818] failed to free device memory at 0x1d8c005800; result: CUDA_ERROR_LAUNCH_TIMEOUT: the launch timed out and was terminated
[taurusi2107:31475] Signal: Aborted (6)
[taurusi2107:31475] Signal code:  (-6)
2020-03-18 16:51:47.705358: E tensorflow/stream_executor/stream.cc:332] Error recording event in stream: Error recording CUDA event: CUDA_ERROR_LAUNCH_TIMEOUT: the launch timed out and was terminated; not marking stream as bad, as the Event object may be at fault. Monitor for further errors.
2020-03-18 16:51:47.705412: E tensorflow/stream_executor/cuda/cuda_event.cc:29] Error polling for event status: failed to query event: CUDA_ERROR_LAUNCH_TIMEOUT: the launch timed out and was terminated
2020-03-18 16:51:47.705421: F tensorflow/core/common_runtime/gpu/gpu_event_mgr.cc:273] Unexpected Event status: 1
srun: error: taurusi2107: task 0: Aborted (core dumped)
srun: error: taurusi2108: task 1: Aborted (core dumped)

```

**Comment by ckkuang** [comment]
CUDA launch timeout usually means the kernel is taking too long to run on GPU which is also used for an active display. GPU driver could kick out the kernel to update the display. Perhaps try updating your GPU driver and cuda version? Reducing the batch size or using a dedicated GPU card may also help.

**Comment by Flamefire** [comment]
I tried this on 2 different kinds of nodes. In both cases I started the script via `srun` (SLURM).

GeForce GTX 1080 Ti:
- Runs fine with AUTO for communication
- Hangs indefinitely when switching to NCCL
- nvidia-smi shows no running processes
- Driver Version: 418.87.01    CUDA Version: 10.1

Tesla K80:
- Runs fine with AUTO
- Shows LAUNCH_TIMEOUT when switching to NCCL
- nvidia-smi shows `/usr/bin/X` and `/usr/bin/gnome-shell` with a single-digit MB usage
- Driver  Version: 418.87.01    CUDA Version: 10.1

Both are only dedicated GPUs, no Display attached.

Reducing the batch size down to 8/worker leads to `Caught signal 11 (Segmentation fault: address not mapped to object at address 0x80)` in `tensorflow::NcclReducer::Run()`.    
Reducing to only 16/worker didn't change anything.

**Comment by dubey** [comment]
Hi, can you clarify whether you can run nccl tests (no TF dependency) across nodes?  Can you also run with `NCCL_DEBUG` environment variable set to `INFO` and paste the logs?  Thanks.

**Comment by Flamefire** [comment]
> can you clarify whether you can run nccl tests (no TF dependency) across nodes

I used the code from https://docs.nvidia.com/deeplearning/sdk/nccl-developer-guide/docs/examples.html "Example 2: One Device per Process or Thread" and sucessfully run it. Compiled with `nvcc example2.c -lnccl -lmpi`. I also used the "Example 3" with 3 devices/process (changed the nDev line). Both show Success.

I run TF with `NCCL_DEBUG=INFO`. For the GeForce GTX 1080 Ti (hang) the output is:
```
taurusa3:27633:27801 [0] NCCL INFO Bootstrap : Using [0]ib0:10.1.148.3<0> [1]ib1:10.2.148.3<0>
taurusa3:27633:27801 [0] NCCL INFO NET/Plugin : No plugin found (libnccl-net.so).
taurusa3:27633:27801 [0] NCCL INFO NET/IB : Using [0]mlx5_1:1/IB [1]mlx5_0:1/IB ; OOB ib0:10.1.148.3<0>
NCCL version 2.4.8+cuda10.1
taurusa5:18071:18468 [0] NCCL INFO Bootstrap : Using [0]ib0:10.1.148.5<0> [1]ib1:10.2.148.5<0>
taurusa5:18071:18468 [0] NCCL INFO NET/Plugin : No plugin found (libnccl-net.so).
taurusa5:18071:18468 [0] NCCL INFO NET/IB : Using [0]mlx5_1:1/IB [1]mlx5_0:1/IB ; OOB ib0:10.1.148.5<0>
taurusa3:27633:28209 [0] NCCL INFO Setting affinity for GPU 0 to 3f
taurusa5:18071:19032 [0] NCCL INFO Setting affinity for GPU 0 to 3f
taurusa3:27633:28210 [1] NCCL INFO Setting affinity for GPU 1 to 0fc0
taurusa5:18071:19033 [1] NCCL INFO Setting affinity for GPU 1 to 0fc0
taurusa3:27633:28212 [2] NCCL INFO Setting affinity for GPU 2 to 0fc0
taurusa5:18071:19034 [2] NCCL INFO Setting affinity for GPU 2 to 0fc0
taurusa5:18071:19032 [0] NCCL INFO CUDA Dev 0[0], IB NIC distance :  PHB PHB
taurusa5:18071:19034 [2] NCCL INFO CUDA Dev 2[2], IB NIC distance :  SYS SYS
taurusa5:18071:19033 [1] NCCL INFO CUDA Dev 1[1], IB NIC distance :  SYS SYS
taurusa3:27633:28209 [0] NCCL INFO CUDA Dev 0[0], IB NIC distance :  PHB PHB
taurusa3:27633:28212 [2] NCCL INFO CUDA Dev 2[2], IB NIC distance :  SYS SYS
taurusa3:27633:28210 [1] NCCL INFO CUDA Dev 1[1], IB NIC distance :  SYS SYS
taurusa3:27633:28209 [0] NCCL INFO Channel 00 :    0   1   2   3   4   5
taurusa3:27633:28209 [0] NCCL INFO Channel 01 :    0   1   2   3   4   5
taurusa3:27633:28209 [0] NCCL INFO Ring 00 : 5 -> 0 [receive] via NET/IB/0
taurusa5:18071:19032 [0] NCCL INFO Ring 00 : 2 -> 3 [receive] via NET/IB/0
taurusa5:18071:19034 [2] NCCL INFO Ring 00 : 5 -> 0 [send] via NET/IB/0
taurusa3:27633:28209 [0] NCCL INFO Ring 00 : 0[0] -> 1[1] via direct shared memory
taurusa5:18071:19032 [0] NCCL INFO Ring 00 : 3[0] -> 4[1] via direct shared memory
taurusa3:27633:28212 [2] NCCL INFO Ring 00 : 2 -> 3 [send] via NET/IB/0
taurusa5:18071:19033 [1] NCCL INFO Ring 00 : 4[1] -> 5[2] via P2P/direct pointer
taurusa3:27633:28210 [1] NCCL INFO Ring 00 : 1[1] -> 2[2] via P2P/direct pointer
taurusa5:18071:19034 [2] NCCL INFO Ring 00 : 5[2] -> 4[1] via P2P/direct pointer
taurusa5:18071:19033 [1] NCCL INFO Ring 00 : 4[1] -> 3[0] via direct shared memory
taurusa3:27633:28212 [2] NCCL INFO Ring 00 : 2[2] -> 1[1] via P2P/direct pointer
taurusa3:27633:28210 [1] NCCL INFO Ring 00 : 1[1] -> 0[0] via direct shared memory
taurusa5:18071:19034 [2] NCCL INFO Ring 01 : 5 -> 0 [send] via NET/IB/1
taurusa3:27633:28212 [2] NCCL INFO Ring 01 : 2 -> 3 [send] via NET/IB/1
taurusa3:27633:28209 [0] NCCL INFO Ring 00 : 3 -> 0 [receive] via NET/IB/0
taurusa5:18071:19032 [0] NCCL INFO Ring 00 : 3 -> 0 [send] via NET/IB/0
taurusa5:18071:19032 [0] NCCL INFO Ring 00 : 0 -> 3 [receive] via NET/IB/0
taurusa3:27633:28210 [1] NCCL INFO Ring 01 : 1[1] -> 2[2] via P2P/direct pointer
taurusa3:27633:28209 [0] NCCL INFO Ring 00 : 0 -> 3 [send] via NET/IB/0
taurusa5:18071:19033 [1] NCCL INFO Ring 01 : 4[1] -> 5[2] via P2P/direct pointer
taurusa3:27633:28209 [0] NCCL INFO Ring 01 : 5 -> 0 [receive] via NET/IB/1
taurusa3:27633:28209 [0] NCCL INFO Ring 01 : 0[0] -> 1[1] via direct shared memory
taurusa5:18071:19032 [0] NCCL INFO Ring 01 : 2 -> 3 [receive] via NET/IB/1
taurusa5:18071:19032 [0] NCCL INFO Ring 01 : 3[0] -> 4[1] via direct shared memory
taurusa5:18071:19034 [2] NCCL INFO Ring 01 : 5[2] -> 4[1] via P2P/direct pointer
taurusa5:18071:19034 [2] NCCL INFO Trees [0] 4->5->-1/-1/-1 [1] 4->5->-1/-1/-1
taurusa3:27633:28212 [2] NCCL INFO Ring 01 : 2[2] -> 1[1] via P2P/direct pointer
taurusa5:18071:19033 [1] NCCL INFO Ring 01 : 4[1] -> 3[0] via direct shared memory
taurusa3:27633:28212 [2] NCCL INFO Trees [0] 1->2->-1/-1/-1 [1] 1->2->-1/-1/-1
taurusa3:27633:28210 [1] NCCL INFO Ring 01 : 1[1] -> 0[0] via direct shared memory
taurusa5:18071:19033 [1] NCCL INFO Trees [0] 3->4->5/-1/-1 [1] 3->4->5/-1/-1
taurusa3:27633:28210 [1] NCCL INFO Trees [0] 0->1->2/-1/-1 [1] 0->1->2/-1/-1
taurusa5:18071:19032 [0] NCCL INFO Ring 01 : 0 -> 3 [receive] via NET/IB/1
taurusa3:27633:28209 [0] NCCL INFO Ring 01 : 0 -> 3 [send] via NET/IB/1
taurusa3:27633:28209 [0] NCCL INFO Ring 01 : 3 -> 0 [receive] via NET/IB/1
taurusa5:18071:19032 [0] NCCL INFO Ring 01 : 3 -> 0 [send] via NET/IB/1
taurusa5:18071:19032 [0] NCCL INFO Trees [0] 0->3->4/-1/-1 [1] -1->3->4/0/-1
taurusa5:18071:19032 [0] NCCL INFO comm 0x2b81240174b0 rank 3 nranks 6 cudaDev 0 nvmlDev 0 - Init COMPLETE
taurusa5:18071:19034 [2] NCCL INFO comm 0x2b811c001430 rank 5 nranks 6 cudaDev 2 nvmlDev 2 - Init COMPLETE
taurusa5:18071:19033 [1] NCCL INFO comm 0x2b8120001190 rank 4 nranks 6 cudaDev 1 nvmlDev 1 - Init COMPLETE
taurusa3:27633:28209 [0] NCCL INFO Trees [0] -1->0->1/3/-1 [1] 3->0->1/-1/-1
taurusa3:27633:28209 [0] NCCL INFO Using 256 threads, Min Comp Cap 6, Trees enabled up to size 69999
taurusa3:27633:28210 [1] NCCL INFO comm 0x2ab7d024cf60 rank 1 nranks 6 cudaDev 1 nvmlDev 1 - Init COMPLETE
taurusa3:27633:28212 [2] NCCL INFO comm 0x2ab7cc050e50 rank 2 nranks 6 cudaDev 2 nvmlDev 2 - Init COMPLETE
taurusa3:27633:28209 [0] NCCL INFO comm 0x2ab7d405e230 rank 0 nranks 6 cudaDev 0 nvmlDev 0 - Init COMPLETE
taurusa3:27633:28021 [0] NCCL INFO Launch mode Group/CGMD
```

And for the K80 (crash):

```
taurusi2107:12777:12926 [0] NCCL INFO Bootstrap : Using [0]ib0:10.1.145.121<0> [1]ib0:0:10.2.145.121<0>
taurusi2107:12777:12926 [0] NCCL INFO NET/Plugin : No plugin found (libnccl-net.so).
taurusi2107:12777:12926 [0] NCCL INFO NET/IB : Using [0]mlx4_0:1/IB ; OOB ib0:10.1.145.121<0>
NCCL version 2.4.8+cuda10.1
taurusi2108:29975:30611 [0] NCCL INFO Bootstrap : Using [0]ib0:10.1.145.122<0> [1]ib0:0:10.2.145.122<0>
taurusi2108:29975:30611 [0] NCCL INFO NET/Plugin : No plugin found (libnccl-net.so).
taurusi2108:29975:30611 [0] NCCL INFO NET/IB : Using [0]mlx4_0:1/IB ; OOB ib0:10.1.145.122<0>
taurusi2107:12777:13231 [0] NCCL INFO Setting affinity for GPU 0 to 0fff
taurusi2108:29975:30730 [0] NCCL INFO Setting affinity for GPU 0 to 0fff
taurusi2108:29975:30731 [1] NCCL INFO Setting affinity for GPU 1 to 0fff
taurusi2107:12777:13245 [1] NCCL INFO Setting affinity for GPU 1 to 0fff
taurusi2108:29975:30733 [3] NCCL INFO CUDA Dev 3[3], IB NIC distance :  PHB
taurusi2108:29975:30730 [0] NCCL INFO CUDA Dev 0[0], IB NIC distance :  SYS
taurusi2108:29975:30732 [2] NCCL INFO CUDA Dev 2[2], IB NIC distance :  PHB
taurusi2108:29975:30731 [1] NCCL INFO CUDA Dev 1[1], IB NIC distance :  SYS
taurusi2107:12777:13231 [0] NCCL INFO CUDA Dev 0[0], IB NIC distance :  SYS
taurusi2107:12777:13246 [2] NCCL INFO CUDA Dev 2[2], IB NIC distance :  PHB
taurusi2107:12777:13245 [1] NCCL INFO CUDA Dev 1[1], IB NIC distance :  SYS
taurusi2107:12777:13247 [3] NCCL INFO CUDA Dev 3[3], IB NIC distance :  PHB
taurusi2107:12777:13231 [0] NCCL INFO Channel 00 :    0   1   6   7   4   5   2   3
taurusi2107:12777:13231 [0] NCCL INFO Channel 01 :    0   1   6   7   4   5   2   3
taurusi2107:12777:13246 [2] NCCL INFO Ring 00 : 5 -> 2 [receive] via NET/IB/0
taurusi2107:12777:13246 [2] NCCL INFO Ring 00 : 2[2] -> 3[3] via P2P/direct pointer
taurusi2108:29975:30732 [2] NCCL INFO Ring 00 : 1 -> 6 [receive] via NET/IB/0
taurusi2108:29975:30733 [3] NCCL INFO Ring 00 : 7[3] -> 4[0] via direct shared memory
taurusi2107:12777:13231 [0] NCCL INFO Ring 00 : 0[0] -> 1[1] via P2P/direct pointer
taurusi2107:12777:13245 [1] NCCL INFO Ring 00 : 1 -> 6 [send] via NET/IB/0
taurusi2107:12777:13247 [3] NCCL INFO Ring 00 : 3[3] -> 0[0] via direct shared memory
taurusi2108:29975:30731 [1] NCCL INFO Ring 00 : 5 -> 2 [send] via NET/IB/0
taurusi2108:29975:30732 [2] NCCL INFO Ring 00 : 6[2] -> 7[3] via P2P/direct pointer
taurusi2108:29975:30730 [0] NCCL INFO Ring 00 : 4[0] -> 5[1] via P2P/direct pointer
taurusi2107:12777:13245 [1] NCCL INFO Ring 00 : 1[1] -> 0[0] via P2P/direct pointer
taurusi2107:12777:13231 [0] NCCL INFO Ring 00 : 0[0] -> 3[3] via direct shared memory
taurusi2108:29975:30730 [0] NCCL INFO Ring 00 : 4[0] -> 7[3] via direct shared memory
taurusi2108:29975:30731 [1] NCCL INFO Ring 00 : 5[1] -> 4[0] via P2P/direct pointer
taurusi2108:29975:30733 [3] NCCL INFO Ring 00 : 7[3] -> 6[2] via P2P/direct pointer
taurusi2107:12777:13247 [3] NCCL INFO Ring 00 : 3[3] -> 2[2] via P2P/direct pointer
taurusi2107:12777:13245 [1] NCCL INFO Ring 01 : 1 -> 6 [send] via NET/IB/0
taurusi2108:29975:30732 [2] NCCL INFO Ring 00 : 6 -> 2 [send] via NET/IB/0
taurusi2107:12777:13246 [2] NCCL INFO Ring 00 : 6 -> 2 [receive] via NET/IB/0
taurusi2108:29975:30731 [1] NCCL INFO Ring 01 : 5 -> 2 [send] via NET/IB/0
taurusi2107:12777:13247 [3] NCCL INFO Ring 01 : 3[3] -> 0[0] via direct shared memory
taurusi2108:29975:30732 [2] NCCL INFO Ring 00 : 2 -> 6 [receive] via NET/IB/0
taurusi2107:12777:13246 [2] NCCL INFO Ring 00 : 2 -> 6 [send] via NET/IB/0
taurusi2107:12777:13231 [0] NCCL INFO Ring 01 : 0[0] -> 1[1] via P2P/direct pointer
taurusi2107:12777:13246 [2] NCCL INFO Ring 01 : 5 -> 2 [receive] via NET/IB/0
taurusi2107:12777:13231 [0] NCCL INFO Ring 01 : 0[0] -> 3[3] via direct shared memory
taurusi2107:12777:13246 [2] NCCL INFO Ring 01 : 2[2] -> 3[3] via P2P/direct pointer
taurusi2107:12777:13246 [2] NCCL INFO Ring 01 : 2 -> 6 [send] via NET/IB/0
taurusi2107:12777:13247 [3] NCCL INFO Ring 01 : 3[3] -> 2[2] via P2P/direct pointer
taurusi2107:12777:13247 [3] NCCL INFO Trees [0] 2->3->0/-1/-1 [1] 2->3->0/-1/-1
taurusi2108:29975:30730 [0] NCCL INFO Ring 01 : 4[0] -> 5[1] via P2P/direct pointer
taurusi2108:29975:30733 [3] NCCL INFO Ring 01 : 7[3] -> 4[0] via direct shared memory
taurusi2108:29975:30731 [1] NCCL INFO Ring 01 : 5[1] -> 4[0] via P2P/direct pointer
taurusi2108:29975:30732 [2] NCCL INFO Ring 01 : 1 -> 6 [receive] via NET/IB/0
taurusi2107:12777:13245 [1] NCCL INFO Ring 01 : 1[1] -> 0[0] via P2P/direct pointer
taurusi2107:12777:13245 [1] NCCL INFO Trees [0] 0->1->-1/-1/-1 [1] 0->1->-1/-1/-1
taurusi2107:12777:13231 [0] NCCL INFO Trees [0] 3->0->1/-1/-1 [1] 3->0->1/-1/-1
taurusi2107:12777:13231 [0] NCCL INFO Using 128 threads, Min Comp Cap 3, Trees enabled up to size 149999
taurusi2108:29975:30732 [2] NCCL INFO Ring 01 : 6[2] -> 7[3] via P2P/direct pointer
taurusi2108:29975:30731 [1] NCCL INFO Trees [0] 4->5->-1/-1/-1 [1] 4->5->-1/-1/-1
taurusi2107:12777:13231 [0] NCCL INFO comm 0x2b6284007b10 rank 0 nranks 8 cudaDev 0 nvmlDev 0 - Init COMPLETE
taurusi2107:12777:13247 [3] NCCL INFO comm 0x2b6240006de0 rank 3 nranks 8 cudaDev 3 nvmlDev 3 - Init COMPLETE
taurusi2108:29975:30730 [0] NCCL INFO Ring 01 : 4[0] -> 7[3] via direct shared memory
taurusi2107:12777:13245 [1] NCCL INFO comm 0x2b6248006de0 rank 1 nranks 8 cudaDev 1 nvmlDev 1 - Init COMPLETE
taurusi2108:29975:30730 [0] NCCL INFO Trees [0] 7->4->5/-1/-1 [1] 7->4->5/-1/-1
taurusi2108:29975:30733 [3] NCCL INFO Ring 01 : 7[3] -> 6[2] via P2P/direct pointer
taurusi2108:29975:30732 [2] NCCL INFO Ring 01 : 2 -> 6 [receive] via NET/IB/0
taurusi2108:29975:30733 [3] NCCL INFO Trees [0] 6->7->4/-1/-1 [1] 6->7->4/-1/-1
taurusi2108:29975:30730 [0] NCCL INFO comm 0x2b6f50007ed0 rank 4 nranks 8 cudaDev 0 nvmlDev 0 - Init COMPLETE
taurusi2108:29975:30731 [1] NCCL INFO comm 0x2b6f4c0071d0 rank 5 nranks 8 cudaDev 1 nvmlDev 1 - Init COMPLETE
taurusi2107:12777:13246 [2] NCCL INFO Ring 01 : 6 -> 2 [receive] via NET/IB/0
taurusi2108:29975:30733 [3] NCCL INFO comm 0x2b6f44007190 rank 7 nranks 8 cudaDev 3 nvmlDev 3 - Init COMPLETE
taurusi2108:29975:30732 [2] NCCL INFO Ring 01 : 6 -> 2 [send] via NET/IB/0
taurusi2108:29975:30732 [2] NCCL INFO Trees [0] 2->6->7/-1/-1 [1] -1->6->7/2/-1
taurusi2107:12777:13246 [2] NCCL INFO Trees [0] -1->2->3/6/-1 [1] 6->2->3/-1/-1
taurusi2108:29975:30732 [2] NCCL INFO comm 0x2b6f480071d0 rank 6 nranks 8 cudaDev 2 nvmlDev 2 - Init COMPLETE
taurusi2107:12777:13246 [2] NCCL INFO comm 0x2b6244006de0 rank 2 nranks 8 cudaDev 2 nvmlDev 2 - Init COMPLETE
taurusi2107:12777:13199 [0] NCCL INFO Launch mode Group
```

Anything else I can do to help?

**Comment by dubey** [comment]
Thanks, looking at the logs above it seems like you are running on 2 hosts each with 1 GPU, right?

Since you are able to run the NCCL tests, this could be a bug in TF.  Can I bother you to run with some more logging enabled?  `TF_CPP_VMODULE=collective_ops=2,nccl_manager=2` should help.

**Comment by Flamefire** [comment]
No, the 1080s are 3/host and the K80 are 4/host (well 2/host but each has 2 GPUs), see `cudaDev 0 - cudaDev 3` in the output above.

Done on the 1080s currently can't get hold of the K80s but can report once done: 
[tfnccl.log](https://github.com/tensorflow/tensorflow/files/4403082/tfnccl.log)

**Comment by Flamefire** [comment]
And here is the log for the K80s with the crash/timeout: 
[tfncclK80.log](https://github.com/tensorflow/tensorflow/files/4407360/tfncclK80.log) Looks like the above so I guess the 1080s simply don't throw a timeout error.

**Comment by sushreebarsa** [comment]
@Flamefire It looks like you are using an older Version of Tensorflow. Many bugs have been fixed in the latest version. Can you please execute your code using Latest Version 2.4.1 or 2.5 and let us know if the issue still persists? Thanks!

**Comment by Flamefire** [comment]
Seems to be fixed in at least 2.4.1
However using NCCL for multi-node is (still?) broken in 2.5.0: https://github.com/tensorflow/tensorflow/issues/50926

Anyway closing this
